wiki/cpp/DEVELOPER_GUIDE.md

```diff
@@ -7,10 +7,43 @@ Please start by reading [CONTRIBUTING.md](../../CONTRIBUTING.md).
 ## Performance
 1. In performance critical sections of the code, favor `cudaDeviceGetAttribute` over `cudaDeviceGetProperties`. See PR [#973](https://github.com/rapidsai/cuml/pull/973) for more details.
 
-## Thread safety
-cuML is thread safe so its functions can be called from multiple host threads if they use different handles.
+## Threading Model
 
-The implementation of cuML is single threaded.
+With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
+assumed to be single threaded. This means they should be able to be called from multiple host threads so 
+long as different instances of `cumlHandle` are used.
+
+Exceptions are made for algorithms that can take advantage of multiple CUDA streams within multiple host threads
+in order to oversubscribe or increase occupancy on a single GPU. In these cases, the use of multiple host 
+threads within cuML algorithms should be used only to maintain concurrency of the underlying CUDA streams. 
+Multiple host threads should be used sparingly, be bounded, and should steer clear of performing CPU-intensive 
+computations.
+
+A good example of an acceptable use of host threads within a cuML algorithm might look like the following
+
+```
+cudaStreamSynchronize(streamInCurrentHostThread);
+
+#pragma omp parallel for num_threads(n_threads)
+for(int i = 0; i < n; i++) {
+    ... possible light cpu pre-processing ...
+    cudaStream_t s;
+    cudaStreamCreate(&s);
+    my_kernel1<<<b, tpb, 0, s>>>(...);
+    ...
+    ... some possible async d2h / h2d copies ...
+    my_kernel2<<<b, tpb, 0, s>>>(...);
+    ...
+    cudaStreamSynchronize(s);
+    ... possible light cpu post-processing ...
+}
+```
+
+To avoid compatibility issues between different threading models, the only threading programming allowed in cuML is OpenMP.
+Though cuML's build enables OpenMP by default, cuML algorithms should still function properly even when OpenMP has been
+disabled. For this reason, it would be better to utilize OpenMP rather than raw pthreads for spawning host threads. 
```

Comment:
Since the only allowed threading model is OpenMP I would remove "For this reason, it would be better to utilize OpenMP rather than raw pthreads for spawning host threads." the reason to only allow OpenMP is already given in the paragraph.